File: re_data_set.rb
#####################################
# Resident Ev-L Ruby interface
#
# The Ruby interface for the 
# Residential Retrofit Evaluator
# Part L(A) dataset
#
####################################
class REDataset
	##
	# Getters
	##
	attr_reader :baseDir, :certsPath, :targetsPath, :retrofitsPath
	##
	# Table file aliases
	##
	CERTS_CSV_NAME		= "certificates"
	TARGETS_CSV_NAME	= "targets"
	RETROFITS_CSV_NAME	= "retrofits"
	##
	# baseDir:	String path to region directory
	##
	def initialize baseDir
		##
		# Throw a fit if the region directory doesn't exist
		##
		unless File.directory? baseDir
			raise "REDataset:BaseDirNotFound"
		end
		##
		# Define stuff
		##
		@baseDir		= baseDir
		@certsPath		= @baseDir + CERTS_CSV_NAME + ".csv"
		@targetsPath	= @baseDir + TARGETS_CSV_NAME + ".csv"
		@retrofitsPath	= @baseDir + RETROFITS_CSV_NAME + ".csv"
		@records		= []
	end
	##
	# idx:		Integer row index
	#
	# Return	RERecord from records
	##
	def [] idx
		@records[idx]
	end
	##
	# Return:	Integer length of table, assuming it's been loaded
	##
	def length
		@records.length
	end
	##
	# Do block with each record
	##
	def each &block
		@records.each{|record| yield record}
	end
	##
	# Remove records which are invalid
	##
	def filterErrors
		return if @errorsRemoved
		@records.select!{|record| ! record.failed}
		@errorsRemoved = true
	end
	##
	# Load Dataset 
	#
	#	Loads the certificates, targets and retrofits tables
	##
	def load 
		# Don't do it twice
		return if @records.any?
		# Pretend to fail gracefully... ... ... ...
		raise "REDataset:DataFileNotFound #{certsPath}" unless File.exists? certsPath
		raise "REDataset:DataFileNotFound #{certsPath}" unless File.exists? targetsPath
		raise "REDataset:DataFileNotFound #{certsPath}" unless File.exists? retrofitsPath
		# Open stuff
		certsFile 			= CSV.open(certsPath)
		targetsFile			= CSV.open(targetsPath)
		retrofitsFile		= CSV.open(retrofitsPath)
		# Get headers
		certsHeaders		= certsFile.shift.map{|header| header.to_sym}
		targetsHeaders		= targetsFile.shift.map{|header| header.to_sym}
		retrofitsHeaders	= retrofitsFile.shift.map{|header| header.to_sym}
		# Read stuff (kind of dirty but whatever)
		while certsRecord = certsFile.shift
			# Get the other 
			targetsRecord	= targetsFile.shift
			retrofitsRecord	= retrofitsFile.shift
			# Build the new record
			newRecord 		= {}
			certsHeaders.each_with_index{|key, idx| newRecord[key] = certsRecord[idx]}
			targetsHeaders.each_with_index{|key, idx| newRecord[key] = targetsRecord[idx]}
			retrofitsHeaders.each_with_index{|key, idx| newRecord[key] = retrofitsRecord[idx]}
			# Append to table
			@records.push RERecord.new(newRecord)
		end
	end
end
##
# Dataset Record, mergeable
##
class RERecord
	##
	# cells:	Hash of record properties
	##
	def initialize cells
		##
		# Convert values into numbers if appropriate
		##
		cells.keys.each{|key|cells[key] = cells[key].to_f rescue cells[key].to_s} 
		@cells = cells
	end
	##
	# Returns:	Bool did the record have errors
	#
	# Note:		Rather that mess with the original certificates
	#			file content, a known identifier from retorfits
	#			is used.
	##
	def failed
		@cells[:"roof-Eff"].to_i == -9999 # As good as anything else, I guess
	end
	##
	# Subscripted access to properties
	##
	def [] key
		@cells[key]
	end
	##
	# Return:	Array of record properties
	##
	def keys
		@keys ||= @cells.keys.dup
	end
end

File: re_data_set.py
#####################################
# Resident Ev-L Python interface
#
# The Python interface for the 
# Residential Retrofit Evaluator
# Part L(A) dataset
#
####################################
# Native
import csv
from os.path import exists
##
# Dataset composite table class 
##
class REDataset():
	##
	# Table file aliases
	##
	CERTS_CSV_NAME		= "certificates.csv"
	TARGETS_CSV_NAME	= "targets.csv"
	RETROFITS_CSV_NAME	= "retrofits.csv"
	##
	# baseDir:	String path to directory with
	# 	the name format "domestic-<government_code>-<County>
	##
	def __init__(self, baseDir):
		##
		# Make sure the dataset exists
		##
		if not exists(baseDir):
			raise "REDataset:BaseDirNotFound - %s" %(baseDir)
		##
		# Define stuff
		##
		self.baseDir 	= baseDir
		# RERecord array
		self.records 	= []
		## Placeholders
		self.loaded		= False
	##
	# Retrieve RERecord by index
	#
	# idx:		Integer index
	#
	# Return:	RERecord
	##
	def __getitem__(self, idx):
		return self.records[idx]
	##
	# Return:	 Integer length of dataset
	##
	def __len__(self):
		return len(self.records)
	##
	# Path to certificates file
	##
	@property
	def certsPath(self):
		return self.baseDir + __class__.CERTS_CSV_NAME 
	##
	# Path to targets file
	##
	@property 
	def targetsPath(self):
		return self.baseDir + __class__.TARGETS_CSV_NAME 
	##
	# Path to retrofits file
	##
	@property
	def retrofitsPath(self):
		return self.baseDir + __class__.RETROFITS_CSV_NAME	
	##
	# Remove records which insufficient data for feature creation
	##
	def filterErrors(self):
		self.records = [record for record in self.records if not record.failed]
	##
	# Load the dataset
	##
	def load(self):
		# Don't do it twice!
		if self.loaded:
			return False
		##
		# Make sure all the files exist
		##
		if not exists(self.certsPath):
			raise "REDataset:CertsFileNotFound"
		if not exists(self.targetsPath):
			raise "REDataset:TargetsFileNotFound"
		if not exists(self.retrofitsPath):
			raise "REDataset:RetrofitsFileNotFound"
		##
		# Open the data files
		##
		certsFile 		= csv.reader(open(self.certsPath))
		targetsFile		= csv.reader(open(self.targetsPath))
		retrofitsFile	= csv.reader(open(self.retrofitsPath))
		##
		# Read records
		## 
		certsHeaders	= next(certsFile)
		targetsHeaders	= next(targetsFile)
		retrofitsHeaders= next(retrofitsFile)
		while True:
			##
			# Read the next row of each file,
			# bail out if you've reached the end
			##
			try:
				certsRecord		= next(certsFile)
			except StopIteration as se:
				break
			targetsRecord	= next(targetsFile)
			retrofitsRecord	= next(retrofitsFile)
			##
			# Build new RERecord input hash
			##
			record			= {}
			# Certificates
			for idx, value in enumerate(certsRecord):
				record[certsHeaders[idx]] 		= certsRecord[idx]
			# Targets 
			for idx, value in enumerate(targetsRecord):
				record[targetsHeaders[idx]] 	= targetsRecord[idx]
			# Retrofits
			for idx, value in enumerate(retrofitsRecord):
				record[retrofitsHeaders[idx]]	= retrofitsRecord[idx]
			##
			# Add new record
			##
			self.records.append(RERecord(record))
		##
		# Make sure we don't load it again (without malicious anyway)! 
		##
		self.loaded = True

##
# Dataset Record Object
##
class RERecord():
	##
	# cells:	Hash of properties
	##
	def __init__(self, cells):
		self.cells	= {}
		for key, value in cells.items():
			try: 
				self.cells[key] = float(value)
			except:
				self.cells[key] = str(value)
	##
	# Retrieve property - subscripted [] access
	#
	# key:	String key of property
	##
	def __getitem__(self, key):
		return self.cells[key]
	##
	# Return:	Array of record property labels
	##
	@property 
	def keys(self):
		return list(self.cells)
	##
	# Returns:	Bool did the record have errors
	#
	# Note:		Rather that mess with the original certificates
	#			file content, a known identifier from retorfits
	#			is used.
	##
	@property
	def failed(self):
		return self.cells["roof-Eff"] == -9999.0 # Since we don't track error in certs
